recipes/parallelproj/build.sh

```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+set -ex
+
+mkdir build
+cd build
+
+if [[ ${cuda_compiler_version:-None} != "None" ]]; then
+  EXTRA_CMAKE_ARGS="-DCMAKE_CUDA_ARCHITECTURES=all"
+else
+  EXTRA_CMAKE_ARGS="-DSKIP_CUDA_LIB=TRUE"
+fi
+
+cmake -G Ninja \
+    ${CMAKE_ARGS} \
+    -DCMAKE_INSTALL_PREFIX=${PREFIX} \
+    -DCMAKE_BUILD_TYPE=Release \
+    -DCMAKE_INSTALL_LIBDIR=lib \
+    -DSKIP_DOCS=TRUE \
+    ${EXTRA_CMAKE_ARGS} \
+    ${SRC_DIR}
+
+cmake --build . --target install --verbose 
```

Comment:
Why are you shipping the source file `projector_kernels.cu.1.2.6` in the lib directory?

Reply:
The reason is because using this file, python users can use cupy's RawKernel to directly execute the kernels on cupy GPU arrays (or also on pytorch tensors).
Is there a better way to ship the file?

Reply:
If the end-use-case is CuPy RawKernel, put this file in python module and load it using `importlib.resources`. [That's what I do.](https://github.com/AdvancedPhotonSource/tike/blob/362463324d904806ab65c4ea425f9835d047f0e3/src/tike/operators/cupy/patch.py#L15) It works a lot better for python users because they don't need to hunt down the file in their system.

Also, the users who would want this source file are probably not the same ones who want the full pre-compiled library?

You could add the CuPy RawKernel Python module as a separate output to this recipe.

Reply:
I think most python users will want both (the pre-compiled cuda lib and the kernel file) such they can either use ctypes to have an interface for the cuda lib for numpy arrays, or the cupy interface for gpu arrays.

I have a separate repository with a python package that uses the compiled libs. And I thought it would be a good idea to keep them separate, since there will be users that only want the compiled c/cuda libs.

Reply:
How does the python project use the raw `.cu` file? Is it JIT-compiled somehow? And is there some usecase on the python side that couldn't be handled by correctly loading the precompiled lib?

Reply:
It is indeed JIT-compiled by cupy which allows the python end user to execute the kernel directly on cupy GPU arrays. The functions in the precompiled lib, take host arrays as input which are then transfered to the GPU where the kernel is executed. After executing the result is transfered back to the host. Using the cupy RawKernel, we can avoid memory transfer between host and GPU which is useful in iterative algorithms where we need to execute kernels many times.

Reply:
Thanks for the explanation. I'd be surprised if this were the first package where this situation arises.

@jakirkham @kkraus14 @jaimergp @isuruf 
Do we have experience/standards for how to distribute raw `.cu` kernels for JIT-compilation?

Reply:
> If the end-use-case is CuPy RawKernel, put this file in python module and load it using `importlib.resources`. [That's what I do.](https://github.com/AdvancedPhotonSource/tike/blob/362463324d904806ab65c4ea425f9835d047f0e3/src/tike/operators/cupy/patch.py#L15) It works a lot better for python users because they don't need to hunt down the file in their system.
> 
> Also, the users who would want this source file are probably not the same ones who want the full pre-compiled library?
> 
> You could add the CuPy RawKernel Python module as a separate output to this recipe.

Putting the .cu kernel file in a python module is one option, but what is users of other high-level languages that support cuda JIT compilation want to use the kernels? For me it seems better to only ship the .cu kernel file independent of the high-level language that wants to use it.

Reply:
> For me, it seems better to only ship the .cu kernel file independent of the high-level language.

+1

Reply:
@carterbox @h-vetinari so shipping the .cu kernel file in the current way is ok?

Reply:
> shipping the .cu kernel file in the current way is ok?

Yes, It's just unusual.

> I have a separate repository with a python package that uses the compiled libs. And I thought it would be a good idea to keep them separate, since there will be users that only want the compiled c/cuda libs.

We prefer that you link that project to the libs in this project dynamically instead of static linking. If the APIs share a version number, then you can build them from the same recipe. If not, then keep them separate.

Reply:
The version numbers are different (I expect that the python version numbers will increase much quicker).
That is because the python package offers an interface to the functions defined in the libs, but also more things.
So we should keep them separated.